### ISSUE TITLE
SMB mount does not work on containerd on Windows

### ISSUE THREAD

**Issue by andyzhangx** [kubernetes/kubernetes#94213]
**What happened**:
Deploy a Windows cluster with containerd configured, SMB mount does not work by trying following project:
 - https://github.com/kubernetes-csi/csi-driver-smb
 - https://github.com/kubernetes-sigs/azurefile-csi-driver

It reports following error:
```

Events:
  Type     Reason            Age                    From                 Message
  ----     ------            ----                   ----                 -------
  Warning  FailedScheduling  4m42s (x3 over 5m1s)   default-scheduler    running "VolumeBinding" filter plugin for pod "busybox-azurefile-0": pod has unbound immediate PersistentVolumeClaims
  Normal   Scheduled         4m40s                  default-scheduler    Successfully assigned default/busybox-azurefile-0 to 1094k8s001
  Normal   Pulling           4m24s                  kubelet, 1094k8s001  Pulling image "e2eteam/busybox:1.29"
  Normal   Pulled            3m55s                  kubelet, 1094k8s001  Successfully pulled image "e2eteam/busybox:1.29"
  Warning  Failed            3m55s                  kubelet, 1094k8s001  Error: failed to generate container "44bea7c01ab600727305945241e1db481be97d6aaf7c3118b527eac86c92aee0" spec: failed to generate spec: failed to resolve symlink "": The system cannot find the file specified.
  Warning  Failed            3m54s                  kubelet, 1094k8s001  Error: failed to generate container "a78a01134aa4d840253501b541e1f31eadac6e9ae29d049883c4cc1a79e45f6e" spec: failed to generate spec: failed to resolve symlink "": The system cannot find the file specified.
  Warning  Failed            3m41s                  kubelet, 1094k8s001  Error: failed to generate container "9686356d344bc1e3c7113277d7e22999cd274d3caa1deba75e813186f987d636" spec: failed to generate spec: failed to resolve symlink "": The system cannot find the file specified.
  Warning  Failed            3m28s                  kubelet, 1094k8s001  Error: failed to generate container "238476ccb56952ddb08dc545d5691538e56bdbb06dfdd800f3cd9f058381065c" spec: failed to generate spec: failed to resolve symlink "": The system cannot find the file specified.
  Warning  Failed            3m14s                  kubelet, 1094k8s001  Error: failed to generate container "d39de41599e788a090604a181d06912b7fc8319ca470f3adc10b2e882fac3862" spec: failed to generate spec: failed to resolve symlink "": The system cannot find the file specified.
  Warning  Failed            2m59s                  kubelet, 1094k8s001  Error: failed to generate container "90e9a6d4187b8381f7c16170c89830404a3c56b235724db96ceee597222f4688" spec: failed to generate spec: failed to resolve symlink "": The system cannot find the file specified.
  Warning  Failed            2m48s                  kubelet, 1094k8s001  Error: failed to generate container "b7cc74cfc596f7774337a5dd2be900f6b33ad2340b37dc86fc259dc62228c31b" spec: failed to generate spec: failed to resolve symlink "": The system cannot find the file specified.
  Warning  Failed            2m36s                  kubelet, 1094k8s001  Error: failed to generate container "92b62f3e6b02b7dc7ffa97c898346bbc821d466314a142fb99dd436743ebc853" spec: failed to generate spec: failed to resolve symlink "": The system cannot find the file specified.
  Warning  Failed            2m24s                  kubelet, 1094k8s001  Error: failed to generate container "893a5d27796c34ae9bfd5a140bc4557aadaf0b3e52c4cca69befcfec6388676f" spec: failed to generate spec: failed to resolve symlink "": The system cannot find the file specified.
  Normal   Pulled            100s (x11 over 3m54s)  kubelet, 1094k8s001  Container image "e2eteam/busybox:1.29" already present on machine
  Warning  Failed            100s (x3 over 2m9s)    kubelet, 1094k8s001  (combined from similar events): Error: failed to generate container "5a7b513e6b4fce31b4968737b573287e4b75a3bf541a7ec9db0880e82c9fcd6c" spec: failed to generate spec: failed to resolve symlink "": The system cannot find the file specified.

I0825 02:11:13.264774    4476 kubelet.go:1953] SyncLoop (PLEG): "busybox-azurefile-0_default(8a6c81a5-761f-4c6f-9cb0-7e2fd5b1c052)", event: &pleg.PodLifecycleEvent{ID:"8a6c81a5-761f-4c6f-9cb0-7e2fd5b1c052", Type:"ContainerStarted", Data:"73e1e4ebb8130d15a1babce1441c72a0fec6262cbc311e944b21ebcf2983b744"}
E0825 02:11:41.599214    4476 remote_runtime.go:200] CreateContainer in sandbox "73e1e4ebb8130d15a1babce1441c72a0fec6262cbc311e944b21ebcf2983b744" from runtime service failed: rpc error: code = Unknown desc = failed to generate container "44bea7c01ab600727305945241e1db481be97d6aaf7c3118b527eac86c92aee0" spec: failed to generate spec: failed to resolve symlink "": The system cannot find the file specified.
E0825 02:11:41.599214    4476 kuberuntime_manager.go:801] container start failed: CreateContainerError: failed to generate container "44bea7c01ab600727305945241e1db481be97d6aaf7c3118b527eac86c92aee0" spec: failed to generate spec: failed to resolve symlink "": The system cannot find the file specified.
E0825 02:11:41.599214    4476 pod_workers.go:191] Error syncing pod 8a6c81a5-761f-4c6f-9cb0-7e2fd5b1c052 ("busybox-azurefile-0_default(8a6c81a5-761f-4c6f-9cb0-7e2fd5b1c052)"), skipping: failed to "StartContainer" for "busybox-azurefile" with CreateContainerError: "failed to generate container \"44bea7c01ab600727305945241e1db481be97d6aaf7c3118b527eac86c92aee0\" spec: failed to generate spec: failed to resolve symlink \"\": The system cannot find the file specified."
E0825 02:11:42.487537    4476 remote_runtime.go:200] CreateContainer in sandbox "73e1e4ebb8130d15a1babce1441c72a0fec6262cbc311e944b21ebcf2983b744" from runtime service failed: rpc error: code = Unknown desc = failed to generate container "a78a01134aa4d840253501b541e1f31eadac6e9ae29d049883c4cc1a79e45f6e" spec: failed to generate spec: failed to resolve symlink "": The system cannot find the file specified.
E0825 02:11:42.487537    4476 kuberuntime_manager.go:801] container start failed: CreateContainerError: failed to generate container "a78a01134aa4d840253501b541e1f31eadac6e9ae29d049883c4cc1a79e45f6e" spec: failed to generate spec: failed to resolve symlink "": The system cannot find the file specified.
E0825 02:11:42.487537    4476 pod_workers.go:191] Error syncing pod 8a6c81a5-761f-4c6f-9cb0-7e2fd5b1c052 ("busybox-azurefile-0_default(8a6c81a5-761f-4c6f-9cb0-7e2fd5b1c052)"), skipping: failed to "StartContainer" for "busybox-azurefile" with CreateContainerError: "failed to generate container \"a78a01134aa4d840253501b541e1f31eadac6e9ae29d049883c4cc1a79e45f6e\" spec: failed to generate spec: failed to resolve symlink \"\": The system cannot find the file specified."
E0825 02:11:55.567146    4476 remote_runtime.go:200] CreateContainer in sandbox "73e1e4ebb8130d15a1babce1441c72a0fec6262cbc311e944b21ebcf2983b744" from runtime service failed: rpc error: code = Unknown desc = failed to generate container "9686356d344bc1e3c7113277d7e22999cd274d3caa1deba75e813186f987d636" spec: failed to generate spec: failed to resolve symlink "": The system cannot find the file specified.
E0825 02:11:55.567146    4476 kuberuntime_manager.go:801] container start failed: CreateContainerError: failed to generate container "9686356d344bc1e3c7113277d7e22999cd274d3caa1deba75e813186f987d636" spec: failed to generate spec: failed to resolve symlink "": The system cannot find the file specified.
E0825 02:11:55.567146    4476 pod_workers.go:191] Error syncing pod 8a6c81a5-761f-4c6f-9cb0-7e2fd5b1c052 ("busybox-azurefile-0_default(8a6c81a5-761f-4c6f-9cb0-7e2fd5b1c052)"), skipping: failed to "StartContainer" for "busybox-azurefile" with CreateContainerError: "failed to generate container \"9686356d344bc1e3c7113277d7e22999cd274d3caa1deba75e813186f987d636\" spec: failed to generate spec: failed to resolve symlink \"\": The system cannot find the file specified."
time="2020-08-25T02:12:03Z" level=error msg="failed to waitForNotification: callbackNumber 5 does not exist in callbackMap"
```

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): v1.18.6
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`): `Microsoft Windows [Version 10.0.17763.1158]`
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

/sig windows
/sig storage

**Comment by marosset** [comment]
I was able to repro you the issue using the azurefile-csi-driver for Windows.

I tried to set up a repo where I would mount an azure file share to a Windows node manually, create symlinks to the share and then schedule a pod to the node using a hostPath volume and the container was able to start up fine using contianerd.

Here is the pod-spec i used
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: host-path-pod
spec:
    nodeSelector:
      "kubernetes.io/os": windows
    containers:
      - name: host-path-pod
        image: e2eteam/busybox:1.29
        command:
          - "powershell.exe"
          - "-Command"
          - "while (1) { Add-Content -Encoding Ascii C:\\mnt\\host-path\\data.txt $(Get-Date -Format u); sleep 10 }"
        volumeMounts:
          - name: host-path-volume
            mountPath: /mnt/host-path
    volumes:
    - name: host-path-volume
      hostPath:
        path: c:/link
        type: Directory
```

I tried to mount the Azure file share with both **New-PsDrive** and **New-SmbGlobalMapping** and saw date being written the the data.txt in the azure file share in both instances.

I used **mklink /D** to create the symlinks

```cmd

PS C:\> Get-SmbGlobalMapping

Status Local Path Remote Path
------ ---------- -----------
OK     X:         \\marossettest.file.core.windows.net\k8stest


c:\> dir c:\
 Volume in drive C is Windows
 Volume Serial Number is 80B5-FDF8

 Directory of c:\

09/01/2020  01:39 AM    <DIR>          AzureData
09/01/2020  12:55 AM    <DIR>          k
08/20/2020  01:41 AM        99,520,042 k.zip
09/01/2020  03:20 AM    <SYMLINKD>     link [c:\t] *
08/20/2020  01:41 AM            77,540 NOTICE.txt
08/20/2020  01:19 AM    <DIR>          Packages
07/10/2020  07:15 PM    <DIR>          PerfLogs
09/01/2020  12:53 AM    <DIR>          Program Files
09/15/2018  07:21 AM    <DIR>          Program Files (x86)
08/20/2020  01:43 AM            12,488 release-notes.txt
09/01/2020  03:20 AM    <SYMLINKD>     t [X:\] *
09/01/2020  01:46 AM    <DIR>          test
09/01/2020  12:53 AM    <DIR>          tmp
09/01/2020  01:45 AM    <DIR>          Users
09/01/2020  12:55 AM    <DIR>          var
08/20/2020  01:41 AM                78 vhd-id.txt
09/01/2020  12:50 AM    <DIR>          Windows
09/01/2020  12:58 AM    <DIR>          WindowsAzure

c:\>dir c:\link
 Volume in drive C is Windows
 Volume Serial Number is 80B5-FDF8

 Directory of c:\link

09/01/2020  03:19 AM    <DIR>          .
09/01/2020  03:19 AM    <DIR>          ..
09/01/2020  03:35 AM             1,914 data.txt
               1 File(s)          1,914 bytes
               2 Dir(s)   1,073,676,288 bytes free

c:\>type c:\link\data.txt
2020-09-01 03:21:14Z
2020-09-01 03:21:24Z
2020-09-01 03:21:34Z
2020-09-01 03:21:44Z
2020-09-01 03:21:54Z
2020-09-01 03:22:04Z
<snip/>
```

Right now I think this issue may be specific to either how the csi-proxy is setting up the SMB share for Windows or how paths are getting passed around between the various components.

**Comment by andyzhangx** [comment]
this is not related to csi-proxy, in-tree azure file does not use csi-proxy, and it does not work

**Comment by andyzhangx** [comment]
hostPath on containerd Windows does not work too: https://github.com/kubernetes/kubernetes/issues/95547

**Comment by marosset** [comment]
I captured containerd logs for this and see

```
Mounts:[]*Mount{
   &Mount{
      ContainerPath:c:/mnt/azurefile,
      HostPath:c:\\var\\lib\\kubelet\\pods\\2d770bfd-ab2b-4157-b0de-42206ef08351\\volumes\\kubernetes.io~csi\\pvc-f43ad1b9-98b7-4704-ac3a-39fc47018771\\mount,
      Readonly:false,
      SelinuxRelabel:false,
      Propagation:PROPAGATION_PRIVATE,},
    &Mount{
      ContainerPath:c:/var/run/secrets/kubernetes.io/serviceaccount,
      HostPath:c:\\var\\lib\\kubelet\\pods\\2d770bfd-ab2b-4157-b0de-42206ef08351\\volumes\\kubernetes.io~secret\\default-token-wt4v2,
      Readonly:true,
      SelinuxRelabel:false,
      Propagation:PROPAGATION_PRIVATE,},
   &Mount{
      ContainerPath:c:/dev/termination-log,
      HostPath:c:\\var\\lib\\kubelet\\pods\\2d770bfd-ab2b-4157-b0de-42206ef08351\\containers\\busybox\\3588f9e0,
      Readonly:false,
      SelinuxRelabel:false,
      Propagation:PROPAGATION_PRIVATE,},},
```

I spoke with @kevpar and he thinks this may be related to a known issue where go-lang is getting confused about on Windows if a symlink points directly to the root of a SBM share (which is does in this case). 
He suggested adding a trailing **\** to see if this fixes the issue.

I will try that out now and report back.

**Comment by marosset** [comment]
I tested some updates to CSI proxy to ensure the symlink to the \\share\path had a trailing backslash and it looks like that resolved the issue.

On my node running contianerd i have the following symlinks 

```
 Directory of c:\var\lib\kubelet\pods\7819fd8b-cdbd-4292-8554-769ad4af543a\volumes\kubernetes.io~csi\pvc-ee3f6a92-f195-4d9e-841b-6966e6839829

10/19/2020  11:11 PM    <DIR>          .
10/19/2020  11:11 PM    <DIR>          ..
10/19/2020  11:11 PM    <SYMLINKD>     mount [c:\var\lib\kubelet\plugins\kubernetes.io\csi\pv\pvc-ee3f6a92-f195-4d9e-841b-6966e6839829\globalmount]
10/19/2020  11:10 PM               340 vol_data.json
               1 File(s)            340 bytes
               3 Dir(s)  117,465,772,032 bytes free

azureuser@1876k8s00000000 c:\k>dir c:\var\lib\kubelet\plugins\kubernetes.io\csi\pv\pvc-ee3f6a92-f195-4d9e-841b-6966e6839829
 Volume in drive C is Windows
 Volume Serial Number is CA78-414E

 Directory of c:\var\lib\kubelet\plugins\kubernetes.io\csi\pv\pvc-ee3f6a92-f195-4d9e-841b-6966e6839829

10/19/2020  11:11 PM    <DIR>          .
10/19/2020  11:11 PM    <DIR>          ..
10/19/2020  11:11 PM    <SYMLINKD>     globalmount [\\faf486191df3c4f0fad548f.file.core.windows.net\pvc-ee3f6a92-f195-4d9e-841b-6966e6839829\]
10/19/2020  11:10 PM               135 vol_data.json
               1 File(s)            135 bytes
               3 Dir(s)  117,465,772,032 bytes free
```

Previously the \ on \\faf486191df3c4f0fad548f.file.core.windows.net\pvc-ee3f6a92-f195-4d9e-841b-6966e6839829\ was missing.

**Comment by kevpar** [comment]
The core issue here is that containerd calls `path/filepath.EvalSymlinks` on the mount source path, and that function has a bug which prevents it from working when the path is a UNC share root. If the path has a trailing `\` appended, then `EvalSymlinks` works properly.

I have filed a bug on Go to fix this, and am working on a fix: golang/go#42079

**Comment by andyzhangx** [comment]
This issue is not resolved, right? @marosset

**Comment by andyzhangx** [comment]
I see this issue is resolved in csi-proxy (only on master branch), while it's not resolved for the in-tree azure file driver.

**Comment by marosset** [comment]
> I see this issue is resolved in csi-proxy v0.2.2, while it's not resolved for the in-tree azure file driver.

Ah, good point.
yes, this is not fully resolved.
